test/smartwallet/CustomSmartWallet.test.ts

```diff
@@ -0,0 +1,339 @@
+
+      const tknBalance = await getTokenBalance(token, worker)
+      const swTknBalance = await getTokenBalance(token, smartWallet.address)
+
+      assert.equal(tknBalance.sub(initialWorkerTokenBalance).toString(), new BN(1).toString(), 'Incorrect new worker token balance')
+      assert.equal(initialSWalletTokenBalance.sub(swTknBalance).toString(), new BN(1).toString(), 'Incorrect new smart wallet token balance')
+
+      assert.equal((await smartWallet.nonce()).toString(), initialNonce.add(new BN(1)).toString(), 'verifyAndCall should increment nonce')
+    })
+
+    it("should call function from custom logic with wallet's address", async () => {
+      // Init smart wallet and
+      const customLogic = await ProxyCustomLogic.new()
+      await smartWallet.initialize(senderAddress, customLogic.address, token.address, worker, '0', '400000', '0x')
+
+      const initialWorkerTokenBalance = await getTokenBalance(token, worker)
+      const initialSWalletTokenBalance = await getTokenBalance(token, smartWallet.address)
+
+      const initialNonce = await smartWallet.nonce()
+
+      const relayRequest = await createRequest({
+        data: recipientFunction,
+        to: recipient.address,
+        nonce: initialNonce.toString(),
+        relayHub: worker,
+        tokenContract: token.address,
+        from: senderAddress
+      }, relayData);
+      const reqData: EIP712TypedData = new TypedRequestData(chainId, smartWallet.address, relayRequest)
+      const sig = signTypedData_v4(senderPrivateKey, { data: reqData })
+      const suffixData = bufferToHex(TypedDataUtils.encodeData(reqData.primaryType, reqData.message, reqData.types).slice((1 + countParams) * 32))
+
+      const result = await smartWallet.execute(domainSeparatorHash, suffixData, relayRequest.request, sig, { from: worker })
+
+      // @ts-ignore
+      assert(containsEvent(customLogic.abi, result.receipt.rawLogs, 'LogicCalled'), 'Should call custom logic')
+
+      // @ts-ignore
+      const logs = await recipient.getPastEvents('TestForwarderMessage')
+      assert.equal(logs.length, 1, 'TestRecipient should emit')
+      assert.equal(logs[0].args.origin, worker, 'test "from" account is the tx.origin')
+      assert.equal(logs[0].args.msgSender, smartWallet.address, 'msg.sender must be the smart wallet address')
+
+      const tknBalance = await getTokenBalance(token, worker)
+      const swTknBalance = await getTokenBalance(token, smartWallet.address)
+
+      assert.equal(tknBalance.sub(initialWorkerTokenBalance).toString(), new BN(1).toString(), 'Incorrect new worker token balance')
+      assert.equal(initialSWalletTokenBalance.sub(swTknBalance).toString(), new BN(1).toString(), 'Incorrect new smart wallet token balance')
+
+      assert.equal((await smartWallet.nonce()).toString(), initialNonce.add(new BN(1)).toString(), 'verifyAndCall should increment nonce')
+    })
+
+    it('should revert if logic revert', async () => {
```

Comment:
sorry, could you please explain what `if logic revert` means?

Reply:
if the logic fails/reverts the transaction

Reply:
thanks!